lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -2980,26 +3016,97 @@ class PrimalGenCloner final
             activeParamIndices));
 
     // Emit the VJP.
-    auto vjpAndVJPIndices = emitAssociatedFunctionReference(
-        context, getBuilder(), getDifferentiationTask(), indices,
-        AutoDiffAssociatedFunctionKind::VJP, getOpValue(ai->getCallee()),
-        /*invoker*/ {ai, synthesis.task}, [&](DifferentiationTask *newTask) {
-          primalGen.schedulePrimalSynthesisIfNeeded(newTask);
-        });
-    if (!vjpAndVJPIndices) {
+    auto loc = ai->getLoc();
+    auto &builder = getBuilder();
+    auto original = getOpValue(ai->getCallee());
+    auto functionSource = original;
+    SILValue vjp;
+    // If functionSource is a @differentiable function, just extract it.
+    auto originalFnTy = original->getType().castTo<SILFunctionType>();
+    if (originalFnTy->isDifferentiable()) {
+      auto paramIndices = originalFnTy->getDifferentiationParameterIndices();
+      for (auto i : indices.parameters->getIndices()) {
+        if (!paramIndices->contains(i)) {
+          context.emitNondifferentiabilityError(original, invoker,
+              diag::autodiff_function_nondiff_parameter_not_differentiable);
+          errorOccurred = true;
+          return;
+        }
+      }
+      vjp = builder.createAutoDiffFunctionExtract(
+          original.getLoc(), AutoDiffFunctionExtractInst::Extractee::VJP,
+          /*differentiationOrder*/ 1, functionSource);
+    }
+
+    // Check and diagnose non-differentiable arguments.
+    for (unsigned paramIndex : range(originalFnTy->getNumParameters())) {
+      if (indices.isWrtParameter(paramIndex) &&
+              !originalFnTy->getParameters()[paramIndex]
+              .getSILStorageType()
+              .isDifferentiable(context.getModule())) {
+        context.emitNondifferentiabilityError(
+            original, invoker, diag::autodiff_nondifferentiable_argument);
+        errorOccurred = true;
+        return;
+      }
+    }
+    // Check and diagnose non-differentiable results.
+    if (!originalFnTy->getResults()[indices.source]
+            .getSILStorageType()
+            .isDifferentiable(context.getModule())) {
+      context.emitNondifferentiabilityError(
+          original, invoker, diag::autodiff_nondifferentiable_result);
       errorOccurred = true;
       return;
     }
-    auto vjp = vjpAndVJPIndices->first;
+    // If VJP has not yet been found, emit an `autodiff_function` instruction
+    // on the remapped original function operand and `autodiff_function_extract`
+    // the VJP. The actual JVP/VJP functions will be populated in the
+    // `autodiff_function` during the transform main loop.
+    if (!vjp) {
+      // FIXME: Handle indirect differentiation invokers. This may require some
+      // redesign: currently, each original function + attribute pair is mapped
+      // only to one invoker.
+      /*
+      DifferentiationInvoker indirect(ai, this->original, attr);
```

Comment:
Note: indirect differentiation invokers are currently not handled post-refactoring. I've disabled the test for now.